Fix linter errors for Storage-ImportExport swagger

leoz-ms opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

leoz-msft committed:
Fix lint errors for Storage-ImportExport swagger

leoz-msft committed:
Fix lint errors for missing descriptions and addmore job properties and update the GetBitLockerKeys response schema

1. Fix lint errors for missing descriptions
2. Add more job properties
3. Update the GetBitLockerKeys response schema

leoz-msft committed:
Remove unused resourceGroupName and update the BitLockerKey response properties to follow camel-case

1. Remove unused resourceGroupName parameter from ListJobs API
2. update the BitLockerKey response properties to follow camel-case

leoz-msft committed:
Fix not-required fields and add 201 response example for CreateJob

1. Fix not-required fields
2. Add 201 response example for CreateJob
3. Fix incorrect single quote character

leoz-msft committed:
Update ShippingInformation properties and remove unnecessary enum values in swagger

1. Update ShippingInformation properties
2. remove unnecessary enum values in swagger

leoz-msft committed:
Fix UpdateJob/CreateJob parameter name in swagger to pass validation

Fix UpdateJob/CreateJob parameter name in swagger to pass validation

leoz-msft committed:
Fix CreateJob response in storage-importexport swagger file

Fix CreateJob response in storage-importexport swagger file

olydis reviewed (changes requested):
Your `operationId`s are a bit confusing: You changed *some* to the `MethodGroup_MethodName` pattern, but not all. Also, those who are changed are not ideal: `ListLocations_list` says "list" twice, I especially wouldn't expect that word on the method group. Then, `GetLocation` has no method group. I'd expect something like `Locations_List` and `Locations_Get`. `ListJobsInSubscription_list` feels backwards, `Jobs_ListBySubscription` feels like the right way.

olydis commented:
@ravbhatnagar Changes to existing Swagger. While renaming/restructuring stuff mostly causes SDK breaking changes (@leoz-ms was anything released for this, yet?), there are also a number of brand new properties.

leoz-ms commented:
@Olydis, yes, some new properties are added for Express Disk Service. These won't be available to general public.

leoz-msft committed:
Normalize operation IDs

Normalize operation IDs

olydis commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -39,7 +39,7 @@
                     "StorageImportExport"
                 ],
-                "operationId": "ListLocations_list",
+                "operationId": "Locations_list",
```
uppercase L

leoz-msft committed:
Fix case and add comments in EDS-related properties

Fix case in Locations_List,
Add comments in EDS-related properties

leoz-ms commented:
@olydis have resolved all comments and passed all checks. Could you please help review?

ravbhatnagar commented:
@leoz-ms  Since these have changes that result in SDK breaking changes, the api-version should be updated for these changes. Does that sound reasonable?

leoz-ms commented:
@ravbhatnagar I'm not sure of what the SDK breaking changes are. From our service standpoint, our service stays the same with only a few optional properties added for supporting more scenarios. And these properties are not available to general public. If those breaking changes come from operation ID change, that's purely client side change. For other changes, I'd say they are fixing the incorrect swagger definition, not introducing anything new. So it does not deserve a server side API-version change, agree?

ravbhatnagar commented:
discussed this offline with @leoz-ms. There are 4 new properties - UserStorageAccountId, UserTenantId, CurrentLocation, ReturnLocation. These are not settable by the user. Only the SPN of another service can set these. Even if the user sets them, the service will ignore them. So for now, I have asked @leoz-ms to remove these from swagger. 
Once these properties have been removed, @olydis  this should be good to merge.
Thanks!

leoz-msft committed:
Remove internal properties

Remove internal properties:
Job properties: userStorageAccountId and userTenantId
DriveStatus properties: currentLocation and returnLocation

leoz-msft committed:
Remove internal properties from samples

Remove internal properties from samples

olydis commented:
@ravbhatnagar Sounds good, @leoz-ms also note that you are making a PR against the public Swagger repo! What goes in here must be available for use by anyone.

olydis commented:
@leoz-ms 
> I'm not sure of what the SDK breaking changes are.

You renamed all operations! So if this was published previously (did you ship SDKs/CLIs based on this?), it has different APIs now, hence breaking customers.

leoz-ms commented:
@olydis Confirmed that all properties are now available to every one. For SDK breaking change, we never shipped SDKs/CLIs based on the previous one, so nobody will be broken.

olydis commented:
@leoz-ms You have more operation groups than you probably anticipated: `Locations`, `Location`, `Jobs`, `Job`, `BitLockerKeys`, `AvailableOperations`.
You probably wanna rename `Job` to `Jobs` and `Location` to `Locations`, so related operations are grouped together. The operation you call `AvailableOperations_List` is usually called `Operations_List`, you may wanna follow the convention.

leoz-msft committed:
update operation Ids to group them properly

update operation Ids to group them properly

leoz-ms commented:
@olydis Renamed operation Ids Locations to Location, Jobs to Job, BitLockerKeys to BitLockerKey, AvailableOperations to Operation.

olydis commented:
@leoz-ms Note that that's essentially *the opposite* of what I asked by making method groups *singular* instead of *plural*. While this is technically valid, it is not our convention. Especially, you won't find any operation in the entire repository that's called [Operation_List](https://github.com/Azure/azure-rest-api-specs/search?utf8=%E2%9C%93&q=Operation_List&type=), it's always `Operations_List`. Please follow the convention 😉

leoz-ms commented:
@olydis That's a weird convention. When the method group is singular, e.g. **Job_List**, you can easily assume a list of jobs will be returned, no second guessing. But when the operation ID is **Jobs_Create**, it confuses users to create multiple jobs while it can only create one.

leoz-ms commented:
@olydis all right. Will rename them to use plural forms.

leoz-msft committed:
Use plural forms for method groups

olydis commented:
I totally agree about the singular/plural confusion - not sure how this became the de-facto standard. 😕

olydis reviewed (commented):
Some more comments regarding SDK quality

leoz-ms commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-        "percentComplete": {
-          "readOnly": true,
-          "type": "integer",
-          "description": "Overall
```
"x-ms-client-name": "ClientTenantId",

olydis replied:
wow, that one must have been fun to figure out 😖

leoz-ms commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-        "percentComplete": {
-          "readOnly": true,
-          "type": "integer",
-          "description": "Overall
```
Add the following for subscriptionId, resourceGroupName and ClientTenantId: "x-ms-parameter-location": "method"

olydis replied:
note that whatever you put this on will be part of every method signature - not sure you always wanna pass in your `subscriptionId`, that one is actually traditionally a client property :) but you got what I meant with the comment, so sorry for the messed up line numbers :(

leoz-ms replied:
Got it. Remove subscriptionId from the method param set, and keep resourceGroupName.

leoz-ms commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-        "percentComplete": {
-          "readOnly": true,
-          "type": "integer",
-          "description": "Overall
```
This GitHub review experience sucks! I don't see which type you are actually referring to. I guess it's DriveBitLockerKey? Anyway, I added null nextLinkName to Locations_List, Operations_List and BitLockerKeys_List.

olydis replied:
commented that on `GetBitLockerKeysResponse`, so it applies to `GetBitLockerKeysResponse`, `LocationsResponse` and `ListOperationsResponse`

leoz-ms replied:
Done

olydis commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-        "percentComplete": {
-          "readOnly": true,
-          "type": "integer",
-          "description": "Overall
```
@leoz-ms applies to `x-ms-client-flatten` specified in `LocationsResponse`, `GetBitLockerKeysResponse`, `DriveBitLockerKey`, `ListOperationsResponse`

leoz-ms replied:
Got it. Removed these.

olydis commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-        "percentComplete": {
-          "readOnly": true,
-          "type": "integer",
-          "description": "Overall
```
your PR removes `required` from the following type definitions: `Operation`, `ReturnAddress`, `ReturnShipping`, `ShippingInformation`, `PackageInfomation`, `DriveStatus` and possibly more (I think it's easiest if you could just do a quick pass over all types and see if something is required - looks as if the removal was by mistake, since the "old" version of the spec had them, the "new" one does not)

leoz-ms replied:
Operation type can only be used in ListOperation response. Making some properties to be required does not provide any information. Removing it makes much more sense.

leoz-ms replied:
Restored other required fields: ReturnAddress, ReturnShipping, ShippingInformation, PackageInfomation, DriveStatus

olydis commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-        "percentComplete": {
-          "readOnly": true,
-          "type": "integer",
-          "description": "Overall
```
GitHub test

olydis replied:
!!!GITHUB TEST!!!

leoz-msft committed:
Resolved comments

1. Added x-ms-pageable
2. Restored required fields
3. Remove unnecessary x-ms-client-flatten
4. Add x-ms-parameter-location as appropriate
5. Add x-ms-client-name for client tenant id

leoz-ms commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1098 @@
-          "description": "Default value is false. Indicates whether the manifest files on the drives should be copied to block blobs.",
-          "default": false
+        "Operation":{
+            "description": "Describes a supported operation by the Storage Import/Export job API.",
```
resolved

olydis commented on specification/storageimportexport/resource-manager/Microsoft.ImportExport/2016-11-01/storageimportexport.json:
```diff
@@ -1,1210 +1,1138 @@
+                },
+                "countryOrRegion": {
+                    "type": "string",
+                    "description": "The c
```
@leoz-ms just to make sure, is this one supposed to be a client parameter instead of a method parameter? :)

olydis replied:
ahhh of course it shows in the wrong spot... I mean the `Accept-Language` parameter that besides `subscriptionId` (where it makes perfect sense) *may* have been left as a client parameter by accident :)

leoz-ms replied:
I don't expect Accept-Language to change across requests just like what subscriptionId, so it makes sense to leave it as client parameter. BTW, I don't find other swaggers defining it, really surprising.

olydis commented:
@leoz-ms Thanks for the changes! You marking some properties as required (awesome! 👍 ) actually stirred up a tiny inconsistency: `ShippingInformation` doesn't *have* the properties you defined as required 😃 Otherwise, I think this is good to go from my perspective. I'll generate and look at the code for this Swagger once the above bug is fixed to do a final review pass.

leoz-msft committed:
Fix incorrect required field names in ShippingInformation

leoz-ms commented:
@olydis All resolved. Could you please take a last look?

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-node

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-python

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-ruby
